gordon/metrics/ffwd.py

```diff
@@ -0,0 +1,138 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2018 Spotify AB
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import asyncio
+import json
+import time
+
+import zope.interface
+
+from gordon import interfaces
+
+
+class FfwdClientProtocol(asyncio.DatagramProtocol):
+    """Protocol for sending one-off messages to ffwd.
+
+    Args:
+        message (bytes): Message for ffwd agent.
+    """
+    def __init__(self, message):
+        self.message = message
+        self.transport = None
+
+    def connection_made(self, transport):
+        self.transport = transport
+        self.transport.sendto(self.message)
+        self.transport.close()
+
+
+class UDPClient:
+    """Client for sending UDP datagrams.
+
+    Args:
+        ip (str): Destination IP address.
+        port (int): Destination port.
+        loop (asyncio.AbstractEventLoop impl): (optional) Event loop.
+    """
+    def __init__(self, ip, port, loop=None):
+        self.ip = ip
+        self.port = port
+        self.loop = loop or asyncio.get_event_loop()
+
+    async def send(self, metric):
+        """Transform metric to JSON bytestring and send to server.
+
+        Args:
+            metric (dict): Complete metric to send as JSON.
+        """
+        message = json.dumps(metric).encode('utf-8')
+        await self.loop.create_datagram_endpoint(
+            lambda: FfwdClientProtocol(message),
+            remote_addr=(self.ip, self.port))
+
+
+@zope.interface.implementer(interfaces.ITimer)
+class FfwdTimer:
+    """ITimer implementation which sends UDP messages to FFWD on completion.
```

Comment:
This (and other implementer) docstrings could be improved; we already know it implemented these interfaces. Maybe something like """Timer metric emitted immediately upon completion."""

Reply:
Yes, if you're looking at the source code it's obvious, but I was thinking for someone just using the library having the implementation note would be nice to have in the docstring, for example so it shows in `help(FfwdTimer)`, since I don't think the zope things show up there. But I can remove it if you think its excessive.

Reply:
The user of ffwdtimer / other interface implementations & providers don't need to care about it implementing an interface though, right? All they should care about is the methods available to them.

Reply:
I guess that makes sense.  Will remove.